openshift/console
# ODC-7546: Fix that the UI doesn't load on non-openshift clusters

## Fixes

https://issues.redhat.com/browse/OCPBUGS-2860, this is almost same PR as #12212 and #13207 :-/

## Analysis / Root cause

When running the console against a non-OpenShift environment, like a pure k8s, k3s, kind, etc. the console UI doesn't show anything.

## Solution Description

1. Show "Namespaces" instead of "Projects" in the Namespace dropdown.
2. Added a new flag that checks that check if `ConsoleQuickStart` CRD is available, otherwise hide the Quick start link from masthead. 
3. Hide the quick search buttons because the quick search doesn't show anything on non-OpenShift clusters at the moment.
4. Added a new redux `k8s` `RESOURCES` `loaded` state because `RESOURCE` `models` are initialized with an empty map and stay empty on a cluster without any CRD.
5. Updated QuickStartDrawer / QuickStartLoader so that they don't get stuck when the QuickStart API call fails.

## Screen recording

todo

## Unit test coverage report

Unchanged because the expectation is that for OpenShift users nothing changed.

## Tests

### Test with KIND           

* Website: https://kind.sigs.k8s.io/
* GitHub: https://github.com/kubernetes-sigs/kind
* CLI download: https://github.com/kubernetes-sigs/kind/releases                                                                                                                                                                              
* `bridge-kind.sh` from https://github.com/jerolimov/openshift/blob/master/scripts/bridge-kind.sh                      

Tested with:

* Docker Client: 24.0.7
* Docker Server: 24.0.7
* Kind: 0.20.0

#### Setup cluster

```
➜  ~ docker ps -a
CONTAINER ID   IMAGE     COMMAND   CREATED   STATUS    PORTS     NAMES

➜  ~ kind create cluster
Creating cluster "kind" ...
 ✓ Ensuring node image (kindest/node:v1.27.3) 🖼 
 ✓ Preparing nodes 📦  
 ✓ Writing configuration 📜 
 ✓ Starting control-plane 🕹️                                                                                           
 ✓ Installing CNI 🔌 
 ✓ Installing StorageClass 💾 
Set kubectl context to "kind-kind"
You can now use your cluster with:

kubectl cluster-info --context kind-kind

Have a nice day! 👋

➜  ~ docker ps -a
CONTAINER ID   IMAGE                  COMMAND                  CREATED          STATUS          PORTS                       NAMES
56f2011db32b   kindest/node:v1.27.3   "/usr/local/bin/entr…"   28 seconds ago   Up 25 seconds   127.0.0.1:35459->6443/tcp   kind-control-plane
```

#### Start the console (without authentification)

`bridge-kind.sh` automatically creates a req. ServiceAccount, ClusterRoleBinding and token.                            


It contains also a hard-coded CONSOLE_DIR to a local console that you might need to update or remove.                  

```
➜  ~ cd console
➜  ~/git/openshift/console ./build.sh

...

➜  ~ bridge-kind.sh
serviceaccount/console-admin created
clusterrolebinding.rbac.authorization.k8s.io/console-admin created

Start console bridge (/home/christoph/git/openshift/console) without authentification!                                 

Using k8s api https://127.0.0.1:35459 with service-account console-admin (namespace kube-system)                       

W1204 15:55:43.667040 2557476 authoptions.go:205] running with AUTHENTICATION DISABLED!                                
I1204 15:55:43.668505 2557476 main.go:625] Binding to 0.0.0.0:9000...
```

Open http://localhost:9000 and test the console

#### Manual tests and notable differences

Before it just shows an empty page, the browser console logs doesn't contain any error.                                

* General:
  * Console loads successfully 🥳
  * ~~Namespace dropdown shows still "Projects" in it~~ fixed in NamespaceDropdown.tsx                                 
  * Namespace dropdown doesn't allow the user to create a new one
  * A lot of errors in the browser console due missing user settings ConfigMap                                         
  * ~~Quick start navigation item is shown also if ConsoleQuickStart CRD isn't found~~ fixed in masthead-toolbar.jsx   
* Administrator perspective
  * Home doesn't show Overview and Projects (might be correct)
  * Operator Hub navigation item was shown, but doesn't show any information (we should maybe hide it in the future?)  
  * Different examples (Create Pod) uses container image that work only on OpenShift (internal image registry)         
* Developer perspective
  * Add page shows Developer Catalog (All Services, Database and Operator Backed) as well as Samples but with any items (we should maybe hide it in the future?)
  * Import from Git isn't shown (due we can only Build with OpenShift BuildConfig and Tekton Pipelines at the moment, it should work with Tekton but I didn't tested this)
  * TODO: Add > Deploy image doesn't show the form (only the title is visible)                                         
  * TODO: Project page is empty
  * Topology is working fine
  * Perspective or add page is currently less useful because most of the add flows doesn't work                        

#### Tear down

```
ctrl+c

➜  ~ kind destroy cluster kind
```
